<!DOCTYPE html><html lang="en"><head><title>Async Await</title><meta name="description" content="is an exciting feature in Javascript that allows you to avoid promise chain hell."><meta name="viewport" content="width=device-width, user-scalable=no"><link rel="stylesheet" href="/styles/app.css"><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12244949-1"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'UA-12244949-1');
</script></head><body><div class="container"><div class="content"><div class="main"><div class="lead"><div class="date">October&nbsp;2017<a href="#" class="hamburger"></a></div><h1>Async Await</h1><h2>is an exciting feature in Javascript that allows you to avoid promise chain hell.</h2></div><div class="blog-content"><p>Remember <em>&quot;callback hell&quot;</em> in Javascript? Remember when we got Promises and everythig got so much better? And now promises chains have become <em>&quot;promise chain hell&quot;</em>. Async await is the next chapter in the saga of dealing with async code in a sane manner in Javascript.</p>
<h3>The problem</h3><p>Take for example a fetch request.</p>
<pre><code class="language-javascript">function getData() {
  return fetch(&quot;http://newfinds.com&quot;, { method: &quot;POST&quot; }) --
    .then(response =&gt; {
      return response.json();
    })
    .catch(err =&gt; {
      console.error(err);
    });
}
// using the function
getData().then(console.log);
</code></pre>
<p>We have a function that returns a promise, which chains a <code>.then</code> off it to also load the <code>.json()</code> response from the result which is also a promise. This is a pretty simple example, but you can see how this would get more and more complex as we chain more and more actions off our promise.</p>
<h3>Async functions to the rescue</h3><pre><code class="language-javascript">async function getData() {
  const response = await fetch(&quot;http://newfinds.com&quot;, { method: &quot;POST&quot; });
  return await response.json();
}
// using the function
getData().then(console.log);
</code></pre>
<p>You can see how much more terse this approach is. Also its much more familier to programmers coming from another language. You do a line of code, and then the next one executes.</p>
<p>The <code>async</code> declaration on the function tells Javascript that you are enabling <code>await</code> inside of it.</p>
<p><strong>Different syntaxes for defining an async function</strong></p>
<pre><code class="language-javascript">// traditional function
async function functionName() {
  return await something();
}

// function defined in an Object
{
  async doThing() {
    return await soemthing();
  }
}

// function as a parameter
doThing(async () =&gt; {
  return await something();
});

// function defined as const
const functionName = async () =&gt; {
  return await soemthing();
};

// or a one liner
const functionName = async () =&gt; await soemthing();
</code></pre>
<p>Async functions <strong>ALWAYS</strong> return a Promise. Even if you are returning something that does not look like a promise it will be wrapped in a <code>Promise.resolve()</code>.</p>
<p><strong>For Example:</strong></p>
<pre><code class="language-javascript">async function doThing() {
  await someAsyncThing();
  return &quot;I did the thing&quot;;
}

doThing().then(console.log);
// I did the thing
</code></pre>
<h3>Handling Errors</h3><p>You may be wondering,</p>
<blockquote>
<p>That&#39;s great, but how do I handle my errors? I don&#39;t see any <code>.catch()</code> inside those async functions..</p>
</blockquote>
<p>Well you can actually <em>still</em> catch the errors if the promise referenced inside the async function rejects. This is handy if you need to do something inside your async function on a falure.</p>
<pre><code class="language-javascript">function rejects() {
  return Promise.reject(&quot;Kaboom&quot;);
}

async function getData() {
  return await rejects().catch(err =&gt; {
    return Promise.reject(&quot;getData caught it&quot;, err);
  });
}
// using the function
getData()
  .then(console.log)
  .catch(err =&gt; {
    console.log(&quot;caught the error&quot;, err);
  });
// caught the error getData caught it
</code></pre>
<p>As you can see, both <em>&quot;catches&quot;</em> were triggered.</p>
<p>More commmonly, like in the case with our fetch example above, you dont want errors thrown, you want to simply reject them. It is much simpler than trying to catch every Promise in your async function.</p>
<p>Say you were making many fetch calls, and you want to reject if any of them fail.</p>
<pre><code class="language-javascript">async function getData() {
  try {
    return [
      await fetch(&quot;http://newfinds.com/foo&quot;),
      await fetch(&quot;http://newfinds.com&quot;)
    ];
  } catch (err) {
    return Promise.reject(err);
  }
}
// using the function
getData()
  .then(console.log)
  .catch(err =&gt; console.log(&quot;caught&quot;, err));
</code></pre>
<h3>Support</h3><p>It was introduced in <strong>Node 7.6</strong>. For browsers its been around even longer.</p>
<pre><code class="language-javascript">{
  chrome: 59,
  edge: 15,
  firefox: 54,
  safari: 10.1,
  node: 7.6,
  ios: 10.3,
  opera: 4,
}
</code></pre>
<p><a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function">âžœ Read more on MDN</a></p></div></div></div><nav id="side-nav"><ul><li><a href="/">Root</a></li><li><a href="/posts/about">About</a></li><li><a href="/morrow_resume.pdf" download="Morrow-Resume.pdf">Resume</a></li><li><h4>Vue</h4><ul><li><a href="/posts/vuex-module-boilerplate">Vuex Module Boilerplate</a></li><li><a href="/posts/vue-route-params">Vue Route Parameters</a></li></ul></li><li><h4>Javascript</h4><ul><li><a href="/posts/javascript-factory-pattern">Javascript Factory Pattern</a></li><li><a href="/posts/protected-methods-in-javascript">Protected Methods In Javascript</a></li><li><a href="/posts/before-and-after-print-hooks">Before &amp; After Hooks for Printing</a></li><li><a href="/posts/javascript-inheritance">Javascript Inheritance</a></li><li><a href="/posts/backbone-singleton">Backbone.js Singleton</a></li><li><a href="/posts/javascript-patterns"> Javascript Patterns</a></li></ul></li><li><h4>ES6</h4><ul><li class="active"><a href="/posts/async-await">Async Await</a></li><li><a href="/posts/es6-in-the-browser">ES6 in the Browser</a></li></ul></li><li><h4>Projects</h4><ul><li><a href="/posts/blissfull-javascript">Bliss.js</a></li></ul></li><li><h4>Ruby</h4><ul><li><a href="/posts/dynamic-bookmarks-in-ruby">Create Dynamic Bookmarklets in Ruby</a></li></ul></li><li><h4>Other</h4><ul><li><a href="/posts/less-tips-and-tricks">Less Tips &amp; Tricks</a></li></ul></li></ul></nav></div><script src="/scripts/app.js"></script><script src="/scripts/prism.js"> </script></body></html>